linux/ibus-keyman/src/engine.c

```diff
@@ -0,0 +1,731 @@
+        }
+
+        if ((state & IBUS_MOD1_MASK) && is_key_pressed(m_display, key_vec, IBUS_Alt_R)) {
+            km_mod_state |= KM_KBP_MODIFIER_RALT;
+        }
+
+        if ((state & IBUS_CONTROL_MASK) && is_key_pressed(m_display, key_vec, IBUS_Control_R)) {
+            km_mod_state |= KM_KBP_MODIFIER_RCTRL;
+        }
+        // ignoring right shift
+        // if ((state & IBUS_SHIFT_MASK) && is_key_pressed(m_display, key_vec, IBUS_Shift_R)) {
+        //     right_modifier_state |= (IBUS_SHIFT_MASK << 8);
+        // }
+        XCloseDisplay(m_display);
+
+        if ((km_mod_state & KM_KBP_MODIFIER_ALT) && !(km_mod_state & KM_KBP_MODIFIER_RALT))
+        {
+            // Left alt so do not pass to keyman
+            return FALSE;
+        }
+        if ((km_mod_state & KM_KBP_MODIFIER_CTRL) && !(km_mod_state & KM_KBP_MODIFIER_RCTRL))
+        {
+            // Left ctrl so do not pass to keyman
+            return FALSE;
+        }
+    }
+    g_message("DAR: ibus_keyman_engine_process_key_event - km_mod_state=%x", km_mod_state);
+    km_kbp_status event_status = km_kbp_process_event(keyman->state,
+                                   keycode_to_vk[keycode], km_mod_state);
+
+    // km_kbp_state_action_items to get action items
+    size_t num_action_items;
+    gchar utf8[12];
+    gint numbytes;
+    const km_kbp_action_item *action_items = km_kbp_state_action_items(keyman->state,
+                                                     &num_action_items);
+
+    for (int i = 0; i < num_action_items; i++)
+    {
+        switch(action_items[i].type)
+        {
+            case KM_KBP_IT_CHAR:
+                numbytes = g_unichar_to_utf8(action_items[i].character, utf8);
+                if (numbytes > 12)
+                {
+                    g_error("g_unichar_to_utf8 overflowing buffer");
+                }
+                if (numbytes)
+                {
+                    utf8[numbytes] = 0;
+                }
+                //g_message("CHAR action");
+                g_message("CHAR action unichar:U+%04x, bytes:%d, string:%s", action_items[i].character, numbytes, utf8);
+                ibus_keyman_engine_commit_string(keyman, utf8);
+                break;
+            case KM_KBP_IT_MARKER:
+                g_message("MARKER action");
+                break;
+            case KM_KBP_IT_ALERT:
+                g_message("ALERT action");
+                break;
+            case KM_KBP_IT_BACK:
+                g_message("BACK action");
+                g_message("DAR: ibus_keyman_engine_process_key_event - client_capabilities=%x, %x", engine->client_capabilities,  IBUS_CAP_SURROUNDING_TEXT);
+
+                if ((engine->client_capabilities & IBUS_CAP_SURROUNDING_TEXT) != 0) {
+                    g_message("deleting surrounding text %ld chars", action_items[i].erased);
+                    ibus_engine_delete_surrounding_text(engine, action_items[i].erased * -1, action_items[i].erased);
```

Comment:
action_item.erased is going away as the processor will return multiple IT_BACK actions.